File: FMI/ANUL II/SEM 2/MPP/ProiectJava/Zboruri/ZboruriPersistence/src/main/java/ro/mpp2024/repository/ZboruriDBRepository.java
package ro.mpp2024.repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp2024.domain.Zbor;
import ro.mpp2024.utils.Constants;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ZboruriDBRepository implements ZboruriRepository {
    private final JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();

    public ZboruriDBRepository(Properties props) {
        logger.info("Initializing ZboruriDBRepository with properties: {} ", props);
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public int size() {
        return 0;
    }

    @Override
    public Zbor findOne(Integer integer) {
        logger.traceEntry("finding zbor with id {}", integer);
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from zboruri where Id = ?")) {
            preparedStatement.setInt(1, integer);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                if (resultSet.next()) {
                    Zbor zbor = createZbor(resultSet);
                    logger.traceExit();
                    return zbor;
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit("No zbor found with id {}", integer);
        return null;
    }

    private Zbor createZbor(ResultSet resultSet) throws SQLException {
        Integer id = resultSet.getInt("Id");
        String destinatie = resultSet.getString("Destinatie");
        LocalDateTime plecare = LocalDateTime.parse(resultSet.getString("Plecare"), Constants.DATE_TIME_FORMATTER_JSON);
        String aeroport = resultSet.getString("Aeroport");
        int nrLocuri = resultSet.getInt("NrLocuri");
        Zbor zbor = new Zbor(aeroport, destinatie, plecare, nrLocuri);
        zbor.setId(id);
        return zbor;
    }

    @Override
    public Iterable<Zbor> findAll() {
        logger.traceEntry();
        List<Zbor> zboruri = new ArrayList<>();
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from zboruri where NrLocuri > 0")) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Zbor zbor = createZbor(resultSet);
                    zboruri.add(zbor);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
        return zboruri;
    }

    @Override
    public void save(Zbor entity) {
    }

    @Override
    public void delete(Integer integer) {
    }

    @Override
    public void update(Integer integer, Zbor entity) {
        logger.traceEntry("updating zbor with id {}", integer);
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("update zboruri set Destinatie=?, Plecare=?, Aeroport=?, NrLocuri=? where Id=?")) {
            preparedStatement.setString(1, entity.getDestinatie());
            preparedStatement.setString(2, entity.getPlecare().format(Constants.DATE_TIME_FORMATTER_JSON));
            preparedStatement.setString(3, entity.getAeroport());
            preparedStatement.setInt(4, entity.getNrLocuri());
            preparedStatement.setInt(5, integer);
            int result = preparedStatement.executeUpdate();
            logger.trace("Updated {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
    }

    @Override
    public List<Zbor> findAllByDestDateAndMinimumSeats(String destination, LocalDateTime date, int min) {
        logger.traceEntry();
        List<Zbor> zboruri = new ArrayList<>();
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from zboruri where Destinatie = ? and Plecare like ? || '%' and NrLocuri >= ?")) {
            preparedStatement.setString(1, destination);
            preparedStatement.setString(2, date.format(Constants.DATE_FORMATTER));
            preparedStatement.setInt(3, min);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Zbor zbor = createZbor(resultSet);
                    zboruri.add(zbor);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
        return zboruri;
    }
}

File: FMI/ANUL II/SEM 2/MPP/ProiectC#/Zboruri/Persistence/ZboruriDBRepository.cs
using System.Data;
using System.Globalization;
using log4net;
using Model;
using Utils;

namespace Persistence
{
    public class ZboruriDbRepository: IZboruriRepository
    {
        private static readonly ILog Logger = LogManager.GetLogger("ZboruriDbRepository");
        private readonly IDictionary<string, string> _props = new SortedList<string, string>();

        public ZboruriDbRepository(IDictionary<string, string> props)
        {
            Logger.Info("Initializing ZboruriDBRepository");
            _props["ConnectionString"] = "URI=" + props["ConnectionString"] + ", Version=3";
        }

        public void Save(Zbor entity)
        {
        }

        public void Update(int id, Zbor update)
        {
            Logger.InfoFormat("updating zbor with id {0}", id);
            IDbConnection con = DbUtils.GetConnection(_props);
            using var comm = con.CreateCommand();
            comm.CommandText =
                "update zboruri set Destinatie=@destinatie, Plecare=@plecare, Aeroport=@aeroport, NrLocuri=@nr_locuri where Id=@id";
            IDbDataParameter paramDest = comm.CreateParameter();
            IDbDataParameter paramPlecare = comm.CreateParameter();
            IDbDataParameter paramAeroport = comm.CreateParameter();
            IDbDataParameter paramNrLocuri = comm.CreateParameter();
            IDbDataParameter paramId = comm.CreateParameter();
            paramDest.ParameterName = "@destinatie";
            paramPlecare.ParameterName = "@plecare";
            paramAeroport.ParameterName = "@aeroport";
            paramNrLocuri.ParameterName = "@nr_locuri";
            paramId.ParameterName = "@id";
            paramDest.Value = update.Destinatie;
            paramPlecare.Value = update.Plecare.ToString(Constants.DateTimeFormatter);
            paramAeroport.Value = update.Aeroport;
            paramNrLocuri.Value = update.NrLocuri;
            paramId.Value = id;
            comm.Parameters.Add(paramDest);
            comm.Parameters.Add(paramPlecare);
            comm.Parameters.Add(paramAeroport);
            comm.Parameters.Add(paramNrLocuri);
            comm.Parameters.Add(paramId);

            int result = comm.ExecuteNonQuery();
            if (result == 0)
            {
                Logger.Error("No flights updated!");
                throw new RepositoryException("No flights updated!");
            }
            else
            {
                Logger.InfoFormat("Updated {0} instances", result);
            }
        }

        public Zbor FindOne(int id)
        {
            Logger.InfoFormat("finding zbor with id {0}", id);
            IDbConnection con = DbUtils.GetConnection(_props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select Id, Destinatie, Plecare, Aeroport, NrLocuri from zboruri where Id=@id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        Zbor zbor = CreateZbor(dataR);
                        Logger.InfoFormat("Exiting findOne with value {0}", zbor);
                        return zbor;
                    }
                }
            }
            Logger.InfoFormat("Exiting findOne with value {0}", null);
            return null;
        }

        private Zbor CreateZbor(IDataReader dataR)
        {
            int idV = dataR.GetInt32(0);
            string destinatie = dataR.GetString(1);
            DateTime plecare = DateTime.ParseExact(dataR.GetString(2), Constants.DateTimeFormatter, CultureInfo.InvariantCulture);
            string aeroport = dataR.GetString(3);
            int nrLocuri = dataR.GetInt32(4);
            Zbor zbor = new Zbor(destinatie, plecare, aeroport, nrLocuri)
            {
                Id = idV
            };
            return zbor;
        }

        public IEnumerable<Zbor> FindAll()
        {
            Logger.InfoFormat("finding flights");
            List<Zbor> zboruri = new List<Zbor>();
            IDbConnection con = DbUtils.GetConnection(_props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select Id, Destinatie, Plecare, Aeroport, NrLocuri from zboruri";

                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        Zbor zbor = CreateZbor(dataR);
                        zboruri.Add(zbor);
                    }
                }
                Logger.InfoFormat("Found {0} instances", zboruri.Count);
            }
            Logger.InfoFormat("Exiting findAllByDestAndDate with value {0}", null);
            return zboruri;
        }

        public void Delete(int id)
        {
        }

        public List<Zbor> FindAllByDestDateAndMinimumSeats(string destination, DateTime data, int min)
        {
            Logger.InfoFormat("finding flights with destination {0} from {1} with minimum {2} seats", destination, data, min);
            List<Zbor> zboruri = new List<Zbor>();
            IDbConnection con = DbUtils.GetConnection(_props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select Id, Destinatie, Plecare, Aeroport, NrLocuri from zboruri where Destinatie=@destinatie and Plecare like @data || '%' and NrLocuri >= @minimum";
                IDbDataParameter paramDest = comm.CreateParameter();
                IDbDataParameter paramData = comm.CreateParameter();
                IDbDataParameter paramLocuri = comm.CreateParameter();
                paramDest.ParameterName = "@destinatie";
                paramData.ParameterName = "@data";
                paramLocuri.ParameterName = "@minimum";
                paramDest.Value = destination;
                paramData.Value = data.ToString(Constants.DateFormatter);
                paramLocuri.Value = min;
                comm.Parameters.Add(paramDest);
                comm.Parameters.Add(paramData);
                comm.Parameters.Add(paramLocuri);

                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        Zbor zbor = CreateZbor(dataR);
                        zboruri.Add(zbor);
                    }
                }
                Logger.InfoFormat("Found {0} instances", zboruri.Count);
            }
            Logger.InfoFormat("Exiting findAllByDestAndDate with value {0}", null);
            return zboruri;
        }
    }
}